### ISSUE TITLE
Snowflake procedure grants  argument_data_types even provided giving error on planning stage "syntax error line 1 at position 82 unexpected 'VARCHAR'."

### ISSUE THREAD

**Issue by systematics78** [Snowflake-Labs/terraform-provider-snowflake#1506]
**Provider Version**
0.56.1

The provider version you are using.
0.56.1
**Terraform Version**
1.3.6
The version of Terraform you were using when the bug was encountered.
ALL
**Describe the bug**
snowflake_procedure_grant deprecated argument, new key arguments data types
while deployinh giving sql compilation error. For example: SHOW GRANTS ON PROCEDURE "CONTROL_DB"."UTILITY"."PRC_UC_SELECT_GRANT_REFINEMENT"(OBJECT_NAME VARCHAR, DEST_ROLE_NAME VARCHAR):VARCHA)
SQL compilation error: syntax error line 1 at position 93 unexpected 'VARCHAR'. Here should only exists data types, not arguments names as well at the end prepended :VARCHA)
snowflake_procedure_grant deprecated argument, new key arguments data types
while deployinh giving sql compilation error. For example: SHOW GRANTS ON PROCEDURE "CONTROL_DB"."UTILITY"."PRC_UC_SELECT_GRANT_REFINEMENT"(OBJECT_NAME VARCHAR, DEST_ROLE_NAME VARCHAR):VARCHA)
SQL compilation error: syntax error line 1 at position 93 unexpected 'VARCHAR'. Here should only exists data types, not arguments names as well at the end prepended :VARCHA)
**Expected behavior**
normal grants
grants to procedure should work properly
**Code samples and commands**
resource snowflake_procedure_grant GRANT_MONITORDB_CREATE_RESOURCE_MONITOR_F1AI_PROC  {
  count                         = var.f1ai_create == "create" ? 1 : 0
  database_name                 = snowflake_database.MONITOR_DB.name
  schema_name                   = snowflake_schema.MONITOR_DBs_Schemas["WAREHOUSE_TAG_CATALOG"].name
  procedure_name                = "CREATE_RESOURCE_MONITOR"
  argument_data_types           = ["string", "string", "string", "string"]

  privilege                     = "USAGE"
  roles                         = [ var.f1ai_admin_role, ]

  with_grant_option = false
  enable_multiple_grants = true
}
Please add code examples and commands that were run to cause the problem.
N/A
**Additional context**
N/A
N/A

**Comment by mnagaa** [comment]
We have the same issue at `function_grant` resource.

This error may come from following updates:
https://github.com/Snowflake-Labs/terraform-provider-snowflake/pull/1502

@systematics78 can you comment out these resources and pass `terraform plan`??
I attempted to comment out but it shows the following:
```
Error: 001003 (42000): SQL compilation error:
syntax error line 1 at position 75 unexpected 'VARCHAR'.

Error: 001003 (42000): SQL compilation error:
syntax error line 1 at position 75 unexpected 'VARCHAR'.
```

**Comment by systematics78** [comment]
I have all Stored procedures and grants to procedures deployed. When it runs terraform plan I'm getting error:
03:53:37  │ Error: 001003 (42000): SQL compilation error:
03:53:37  │ syntax error line 1 at position 93 unexpected 'VARCHAR'.
03:53:37  │ 
03:53:37  │   with snowflake_procedure_grant.GRANT_USAGE_ON_PRC_CREATE_SNOWFLAKE_SHARE_PRE_TO_PLA,
03:53:37  │   on snowflake_share.tf line 149, in resource "snowflake_procedure_grant" "GRANT_USAGE_ON_PRC_CREATE_SNOWFLAKE_SHARE_PRE_TO_PLA":
03:53:37  │  149: resource snowflake_procedure_grant GRANT_USAGE_ON_PRC_CREATE_SNOWFLAKE_SHARE_PRE_TO_PLA  {

This is because it tries to execute sql statement like show grants to procedure()

actual sql statement should be show grants to procedure (argument_data_types), like for example:
SHOW GRANTS ON PROCEDURE "CONTROL_DB"."UTILITY"."PRC_UC_SELECT_GRANT_REFINEMENT"(VARCHAR, VARCHAR);

but instead of that it executes:
SHOW GRANTS ON PROCEDURE "CONTROL_DB"."UTILITY"."PRC_UC_SELECT_GRANT_REFINEMENT"(OBJECT_NAME VARCHAR, DEST_ROLE_NAME VARCHAR):VARCHA) which totally wrong.

**Comment by mnagaa** [comment]
@systematics78 
Okey. I found generated query by provider is wrong after v.0.56.

**Comment by sfc-gh-swinkler** [comment]
Okay so i think i figured it out. The problem was that prior to 0.56.1 return_type was required and actually used to set in the id. See here:
https://github.com/Snowflake-Labs/terraform-provider-snowflake/blob/8c31439253d25aafb54fc09d89e547fa8238258c/pkg/resources/grant_helpers.go#L475

Return type was removed since it was not actually needed. However, if you created a resource already then your ID still had return_type. Example: `MY_DATABASE|MY_SCHEMA|test_proc(string, string, string, string):varchar|USAGE|MY_ROLE|false` The fix is to simply trim off the return type if it is detected.

**Comment by systematics78** [comment]
Much appreciate, I have 8 accounts in different regions for d/q/p and DR, bunch of stored procedures, so manually managing stages by removing grants for procedure will not work for me.

**Comment by systematics78** [comment]
@sfc-gh-swinkler
One more thing, in query history i can see query like:
SHOW GRANTS ON PROCEDURE "CONTROL_DB"."UTILITY"."PRC_UC_SELECT_GRANT_REFINEMENT"(OBJECT_NAME VARCHAR, DEST_ROLE_NAME VARCHAR):VARCHA)

Here it also using argument name and data type, but it should be only data type coma separated
BR,

**Comment by sfc-gh-swinkler** [comment]
@systematics78 the improper query is the result of the Read function deciphering the old ID incorrectly. 
GrantID read here:
https://github.com/Snowflake-Labs/terraform-provider-snowflake/blob/main/pkg/resources/procedure_grant.go#L203-L204
and parsed here
https://github.com/Snowflake-Labs/terraform-provider-snowflake/blob/main/pkg/resources/procedure_grant.go#L226
The argumentDataTypes was including return_type. So latest PR should fix that, since i specifically check for it and trim it off.

**Comment by sfc-gh-swinkler** [comment]
This was released as part of 0.56.2. Please let me know if there are any further issues.
